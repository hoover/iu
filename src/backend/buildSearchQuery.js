import { DateTime } from 'luxon'
import {
    DEFAULT_FACET_SIZE,
    DEFAULT_INTERVAL,
    DEFAULT_OPERATOR,
    HIGHLIGHT_SETTINGS,
    PRIVATE_FIELDS,
} from '../constants/general'
import { daysInMonth } from '../utils'
import { aggregationFields } from '../constants/aggregationFields'

const expandPrivate = (field, uuid) => {
    if (PRIVATE_FIELDS.includes(field)) {
        return `${field}.${uuid}`
    }
    return field
}

const buildQuery = (q, filters, searchFields) => {
    const qs = {
        query_string: {
            query: q,
            default_operator: DEFAULT_OPERATOR,
            fields: searchFields.all,
            lenient: true,
        },
    }

    const ranges = [];
    ['date', 'date-created'].forEach(field => {
        const value = filters[field]
        if (value?.from && value?.to) {
            ranges.push({
                range: {
                    [field]: {
                        gte: value.from,
                        lte: value.to,
                    },
                },
            })
        }
    })

    if (ranges.length) {
        return {
            bool: {
                must: [qs, ...ranges],
            },
        }
    }

    return qs
}

const buildSortQuery = order => order?.reverse().map(([field, direction = 'asc']) => field.startsWith('_') ?
    {[field]: {order: direction}} :
    {[field]: {order: direction, missing: '_last'}}
) || []

const buildTermsField = (field, uuid, terms, page = 1, size = DEFAULT_FACET_SIZE) => {
    const fieldKey = expandPrivate(field, uuid)

    let filterClause = null
    if (terms?.include?.length) {
        if (aggregationFields[field].type === 'term-and') {
            filterClause = terms.include.map(term => ({
                term: {[fieldKey]: term}
            }))
        } else {
            filterClause = {
                terms: {[fieldKey]: terms.include},
            }
        }
    }

    let filterExclude = null
    if (terms?.exclude?.length) {
        filterExclude = {
            terms: { [fieldKey]: terms.exclude },
        }
    }

    let filterMissing = null
    if (terms?.missing === 'true') {
        filterMissing = false
    } else if (terms?.missing === 'false') {
        filterMissing = true
    }

    if (field === 'tags' && !terms?.include?.includes('trash') && !terms?.exclude?.includes('trash')) {
        if (filterExclude === null) {
            filterExclude = { terms: { [fieldKey]: ['trash'] } }
        } else {
            filterExclude.terms[fieldKey].push('trash')
        }
    }

    return {
        field,
        originalField: field,
        aggregation: {
            terms: {
                field: fieldKey,
                size: page * size,
            },
        },
        filterClause,
        filterExclude,
        filterMissing,
    }
}

const intervalFormat = (interval, param) => {
    switch (interval) {
        case 'year':
            return {
                gte: `${param}-01-01T00:00:00.000Z`,
                lte: `${param}-12-31T23:59:59.999Z`,
            }

        case 'month':
            return {
                gte: `${param}-01T00:00:00.000Z`,
                lte: `${param}-${daysInMonth(param)}T23:59:59.999Z`,
            }

        case 'week':
            return {
                gte: `${param}T00:00:00.000Z`,
                lt: `${DateTime.fromISO(param).plus({days: 7}).toISODate()}T23:59:59.999Z`,
            }

        case 'day':
            return {
                gte: `${param}T00:00:00.000Z`,
                lte: `${param}T23:59:59.999Z`,
            }

        case 'hour':
            return {
                gte: `${param}:00:00.000Z`,
                lte: `${param}:59:59.999Z`,
            }
    }
}

const buildHistogramField = (field, uuid, { interval = DEFAULT_INTERVAL, intervals } = {},
                             page = 1, size = DEFAULT_FACET_SIZE) => {

    const fieldKey = expandPrivate(field, uuid)

    let filterClause = null
    if (intervals?.include?.length) {
        filterClause = {
            bool: {
                should: intervals.include.map(selected => ({
                    range: {
                        [fieldKey]: intervalFormat(interval, selected),
                    },
                })),
            },
        }
    }

    let filterMissing = null
    if (intervals?.missing === 'true') {
        filterMissing = false
    } else if (intervals?.missing === 'false') {
        filterMissing = true
    }

    return {
        field,
        originalField: field,
        aggregation: {
            date_histogram: {
                field: fieldKey,
                interval,
                min_doc_count: 1,
                order: { '_key': 'desc' },
            },
            aggs: {
                bucket_truncate: {
                    bucket_sort: {
                        from: (page - 1) * size,
                        size
                    }
                }
            }
        },
        filterClause,
        filterMissing,
    }
}

const buildMissingField = (field, uuid) => ({
    field: `${field}-missing`,
    originalField: field,
    aggregation: {
        missing: { field: expandPrivate(field, uuid) },
    }
})

const prepareFilter = field => {
    const filter = []

    if (field.filterClause) {
        filter.push(field.filterClause)
    }

    if (field.filterExclude) {
        filter.push({ bool: { must_not: field.filterExclude } })
    }

    if (typeof field.filterMissing === 'boolean') {
        const exists = { exists: { field: field.field } }
        if (field.filterMissing) {
            filter.push(exists)
        } else {
            filter.push({ bool: { must_not: exists } })
        }
    }

    if (filter.length) {
        return filter
    }
}

const buildFilter = fields => {
    const filter = []

    fields.forEach(field => {
        const fieldFilter = prepareFilter(field)

        if (fieldFilter) {
            if (aggregationFields[field.originalField].type === 'term-and') {
                filter.push(...fieldFilter)
            } else {
                filter.push({ bool: { should: fieldFilter } })
            }
        }
    })

    if (filter.length) {
        return { bool: { filter } }
    } else {
        return { bool: {} }
    }
}

const buildAggs = fields => fields.reduce((result, field) => ({
    ...result,
    [field.field]: {
        aggs: {
            values: field.aggregation,
            count: { cardinality: { field: field.field } },
        },
        filter: buildFilter(fields.filter(other => other.originalField !== field.originalField)),
    },
}), {})

const buildSearchQuery = (
    {
        q = '*',
        page = 1,
        size = 0,
        order,
        collections = [],
        facets = {},
        filters = {},
    } = {},
    type,
    searchFields,
    uuid
) => {

    const query = buildQuery(q, filters, searchFields)
    const sort = buildSortQuery(order)

    const dateFields = Object.entries(aggregationFields)
        .filter(([,field]) => field.type === 'date')
        .map(([key]) => key)

    const termFields = Object.entries(aggregationFields)
        .filter(([,field]) => field.type.startsWith('term'))
        .map(([key]) => key)

    const fields = [
        ...dateFields.map(field => buildHistogramField(field, uuid, filters[field], facets[field])),
        ...dateFields.map(field => buildMissingField(field, uuid)),
        ...termFields.map(field => buildTermsField(field, uuid, filters[field], facets[field])),
        ...termFields.map(field => buildMissingField(field, uuid)),
    ]

    const postFilter = buildFilter(fields);
    const aggs = buildAggs(fields);

    const highlightFields = {}
    searchFields.highlight.forEach(field => {
        highlightFields[field] = HIGHLIGHT_SETTINGS
    })

    return {
        from: (page - 1) * size,
        size: type === 'aggregations' ? 0 : size,
        query,
        sort,
        post_filter: postFilter,
        aggs: type === 'results' ? {} : aggs,
        collections,
        _source: searchFields._source,
        highlight: {
            fields: highlightFields,
        },
    }
}

export default buildSearchQuery
