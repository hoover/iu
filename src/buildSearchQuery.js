import { DEFAULT_FACET_SIZE, SORTABLE_FIELDS } from './constants'

function buildQuery(q, { dateCreatedRange, dateModifiedRange }) {
    const qs = {
        query_string: {
            query: q,
            default_operator: 'AND',
        },
    }

    const queryRanges = []
    if (dateCreatedRange) {
        queryRanges.push(['date-created', dateCreatedRange])
    }
    if (dateModifiedRange) {
        queryRanges.push(['date', dateModifiedRange])
    }

    const ranges = []
    queryRanges.forEach(([field, dateRange]) => {
        if (dateRange && dateRange.from && dateRange.to) {
            ranges.push({
                range: {
                    [field]: {
                        gte: dateRange.from,
                        lte: dateRange.to,
                    },
                },
            })
        }
    })

    if (ranges.length) {
        return {
            bool: {
                must: [qs, ...ranges],
            },
        }
    }

    return qs
}

function buildSortQuery(order) {
    let sort = ['_score', '_id'];

    Array.isArray(order) && order.filter(([field, direction = 'asc']) =>
        SORTABLE_FIELDS.includes(field) && ['asc', 'desc'].includes(direction)
    ).reverse().forEach(([field, direction = 'asc']) => {
        sort = [{ [field]: { order: direction, missing: '_last' } }, ...sort]
    })

    return sort;
}

function buildTermsField(
    name,
    terms,
    { field = name, size = DEFAULT_FACET_SIZE } = {}
) {
    return {
        name,
        field,
        aggregation: {
            terms: { field, size },
        },
        filterClause:
            terms && terms.length
                ? {
                      terms: { [field]: terms },
                  }
                : null,
    };
}

function buildYearField(name, value, { field = name } = {}) {
    return {
        name,
        field,
        aggregation: {
            date_histogram: {
                field,
                interval: 'year',
            },
        },
        filterClause:
            value && value.length
                ? {
                      bool: {
                          should: value.map(year => ({
                              range: {
                                  date: {
                                      gte: `${year}-01-01`,
                                      lte: `${year}-12-31`,
                                  },
                              },
                          })),
                      },
                  }
                : null,
    };
}

function buildFilter(fields) {
    const must = fields.map(f => f.filterClause).filter(Boolean);

    if (must.length) {
        return {
            bool: {
                must,
            },
        };
    } else {
        return { bool: {} };
    }
}

function buildAggs(fields) {
    return fields.reduce(
        (result, field) => ({
            ...result,
            [`count_by_${field.name}`]: {
                aggs: {
                    [field.name]: field.aggregation,
                    [`${field.name}_count`]: { cardinality: { field: field.field } },
                },
                filter: buildFilter(
                    fields.filter(other => other.name !== field.name)
                ),
            },
        }),
        {}
    );
}

export default function buildSearchQuery({
    page = 1,
    size = 0,
    q = '*',
    order = null,
    collections = [],
    dateYears = null,
    dateCreatedYears = null,
    dateCreatedRange,
    dateModifiedRange,
    searchAfter = '',
    fileType = null,
    language = null,
    emailDomains = null,
    facets = {},
} = {}) {
    const query = buildQuery(q, { dateCreatedRange, dateModifiedRange });
    const sort = buildSortQuery(order);

    const fields = [
        buildTermsField('filetype', fileType, { size: facets.fileType }),
        buildTermsField('email_domains', emailDomains, {
            field: 'email-domains',
            size: facets.emailDomains,
        }),
        buildTermsField('lang', language, { size: facets.language }),
        buildYearField('date_years', dateYears, {
            field: 'date',
            size: facets.dateYears,
        }),
        buildYearField('date_created_years', dateCreatedYears, {
            field: 'date-created',
            size: facets.dateCreatedYears,
        }),
    ];

    const postFilter = buildFilter(fields);
    const aggs = buildAggs(fields);

    return {
        from: (page - 1) * size,
        size: size,
        query,
        search_after: searchAfter,
        sort,
        post_filter: postFilter,
        aggs,
        collections: collections,
        _source: [
            'path',
            'url',
            'mime_type',
            'attachments',
            'filename',
            'word-count',
            'date',
            'date-created',
        ],
        highlight: {
            fields: {
                '*': {
                    fragment_size: 150,
                    number_of_fragments: 3,
                    require_field_match: false,
                    pre_tags: ['<mark>'],
                    post_tags: ['</mark>'],
                },
            },
        },
    };
}
