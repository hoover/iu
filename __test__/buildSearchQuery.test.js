import buildSearchQuery from '../src/buildSearchQuery';

it('builds a default query', () => {
    const query = buildSearchQuery();
    expect(query).toMatchSnapshot();
});

it('builds a query with a filetype filter', () => {
    const query = buildSearchQuery({ fileType: ['email', 'pdf'] });

    expect(query.post_filter).toMatchObject({
        bool: {
            must: [{ terms: { filetype: ['email', 'pdf'] } }],
        },
    });

    expect(query.aggs.count_by_email_domains).toMatchObject({
        aggs: {
            email_domains: {
                terms: { field: 'email-domains' },
            },
        },
        filter: {
            bool: {
                must: [{ terms: { filetype: ['email', 'pdf'] } }],
            },
        },
    });

    expect(query.aggs.count_by_filetype).toMatchObject({
        aggs: {
            filetype: {
                terms: { field: 'filetype' },
            },
        },
    });
});

it('builds a query with a date year filter', () => {
    const query = buildSearchQuery({ dateYears: ['2009'] });

    const yearFilter = {
        bool: {
            should: [
                {
                    range: {
                        date: {
                            gte: '2009-01-01',
                            lte: '2009-12-31',
                        },
                    },
                },
            ],
        },
    };

    expect(query.post_filter).toMatchObject({
        bool: {
            must: [yearFilter],
        },
    });

    expect(query.aggs.count_by_filetype).toMatchObject({
        aggs: {
            filetype: {
                terms: { field: 'filetype' },
            },
        },
        filter: {
            bool: {
                must: [yearFilter],
            },
        },
    });

    expect(query.aggs.count_by_date_years).toMatchObject({
        aggs: {
            date_years: {
                date_histogram: { field: 'date', interval: 'year' },
            },
        },
        filter: {
            bool: {},
        },
    });
});

it('builds a query with multiple fields filtered', () => {
    const query = buildSearchQuery({
        fileType: ['doc', 'email'],
        emailDomains: ['gmail.com'],
    });

    expect(query.post_filter).toMatchObject({
        bool: {
            must: [
                { terms: { filetype: ['doc', 'email'] } },
                { terms: { 'email-domains': ['gmail.com'] } },
            ],
        },
    });

    expect(query.aggs.count_by_filetype).toMatchObject({
        aggs: {
            filetype: {
                terms: { field: 'filetype' },
            },
        },
        filter: {
            bool: {
                must: [{ terms: { 'email-domains': ['gmail.com'] } }],
            },
        },
    });

    expect(query.aggs.count_by_email_domains).toMatchObject({
        aggs: {
            email_domains: {
                terms: { field: 'email-domains' },
            },
        },
        filter: {
            bool: {
                must: [{ terms: { filetype: ['doc', 'email'] } }],
            },
        },
    });
});
