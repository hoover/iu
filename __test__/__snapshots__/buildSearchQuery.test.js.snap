// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a default query 1`] = `
Object {
  "_source": Array [],
  "aggs": Object {
    "date": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "date",
          },
        },
        "values": Object {
          "aggs": Object {
            "bucket_truncate": Object {
              "bucket_sort": Object {
                "from": 0,
                "size": 100,
              },
            },
          },
          "date_histogram": Object {
            "field": "date",
            "interval": "year",
            "min_doc_count": 1,
            "order": Object {
              "_key": "desc",
            },
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "date-created": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "date-created",
          },
        },
        "values": Object {
          "aggs": Object {
            "bucket_truncate": Object {
              "bucket_sort": Object {
                "from": 0,
                "size": 100,
              },
            },
          },
          "date_histogram": Object {
            "field": "date-created",
            "interval": "year",
            "min_doc_count": 1,
            "order": Object {
              "_key": "desc",
            },
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "date-created-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "date-created-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "date-created",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "date-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "date-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "date",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "email-domains": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "email-domains",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "email-domains",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "email-domains-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "email-domains-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "email-domains",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "filetype": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "filetype",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "filetype",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "filetype-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "filetype-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "filetype",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "from.keyword": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "from.keyword",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "from.keyword",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "from.keyword-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "from.keyword-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "from.keyword",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "lang": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "lang",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "lang",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "lang-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "lang-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "lang",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "path-parts": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "path-parts",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "path-parts",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "path-parts-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "path-parts-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "path-parts",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "priv-tags": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "priv-tags",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "priv-tags.testuser",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "priv-tags-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "priv-tags-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "priv-tags.testuser",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "tags": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "tags",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "tags",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {},
      },
    },
    "tags-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "tags-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "tags",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "to.keyword": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "to.keyword",
          },
        },
        "values": Object {
          "terms": Object {
            "field": "to.keyword",
            "size": 100,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
    "to.keyword-missing": Object {
      "aggs": Object {
        "count": Object {
          "cardinality": Object {
            "field": "to.keyword-missing",
          },
        },
        "values": Object {
          "missing": Object {
            "field": "to.keyword",
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "filter": Array [],
          "must_not": Array [
            Object {
              "terms": Object {
                "tags": Array [
                  "trash",
                ],
              },
            },
          ],
          "should": Array [],
        },
      },
    },
  },
  "collections": Array [],
  "from": 0,
  "highlight": Object {
    "fields": Object {},
  },
  "post_filter": Object {
    "bool": Object {
      "filter": Array [],
      "must_not": Array [
        Object {
          "terms": Object {
            "tags": Array [
              "trash",
            ],
          },
        },
      ],
      "should": Array [],
    },
  },
  "query": Object {
    "query_string": Object {
      "default_operator": "AND",
      "fields": Array [],
      "lenient": true,
      "query": "*",
    },
  },
  "size": 0,
  "sort": Array [],
}
`;
